bazaarvoice/emodb
# Enabled Scanner/Stash to run in separate manager/worker modes

## Github Issue #

None

## What Are We Doing Here?

Stash relies on there always being at least one server running.  This server ensures that the daily Stash starts at the scheduled time and responds to ad-hoc requests as needed.

Currently all Stash servers are equivalent, meaning that the system can scale down to one instance but that instance must be of an instance type which can handle active Stash resource utilization, even though majority of the time it is not being utilized.

This PR optionally allows the administrator to deploy two Stash instance types: a manager and a worker.  The manager is responsible for starting scheduled Stashes, publishing broad Stash metrics and managing the workflow for an active Stash job.  A worker is responsible for performing the actual scanning and uploading of data for an active Stash job.  The advantages of this approach are:

1. The single always-on manager instance for Stash can be a significantly lower-resource and lower-cost type than the worker instances.
2. The more expenseive worker instances can scale up as Stashes are active then be completely torn down when Stash completes.

It is also possible to start a worker which is both a manager and worker, which would work like the current Stash does.

## How to Test and Verify

This is difficult to test locally since by nature the scaling aspects of the service require running on EC2 instances.  However, the logic can be verified as follows:

1. Check out this PR
2. Start EmoDB normally using `web-local/start.sh` and create at least one table.
3. From the `web-local` directory start a stash manager:
```
java -jar ../web/target/emodb-web-5.6.3-SNAPSHOT.jar server ./config-stashmanagertest.yaml ./config-ddl-local.yaml
```
4.  Again from `web-local` start a stash worker:
```
java -jar ../web/target/emodb-web-5.6.3-SNAPSHOT.jar server ./config-stashworkertest.yaml ./config-ddl-local.yaml
```
5.  Make a request to start a Stash job:
```
 curl 'localhost:8780/scanner/1/upload/test-stash?dest=null&placement=ugc_global:ugc' -XPOST
```
6. Look at the logs on each of the stash services and verify the manager only queues range scans while the worker is the one claiming, scanning, and releasing those scans.

## Risk

This is a fairly low risk change.  So long as the services are started as expected the actual work of generating Stash is unchanged.  The only change is where the specific parts of the Stash workflow will run.

### Level 

`Medium`

### Required Testing

`Manual`

Add one or a few complete sentences about the possible risks or concerns for
this change.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
